flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -545,14 +545,12 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
                             .thenApplyAsync(
                                     plan -> {
                                         try {
-                                            CheckpointIdAndStorageLocation
-                                                    checkpointIdAndStorageLocation =
-                                                            initializeCheckpoint(
-                                                                    request.props,
-                                                                    request.externalSavepointLocation,
-                                                                    initializeBaseLocations);
-                                            return new Tuple2<>(
-                                                    plan, checkpointIdAndStorageLocation);
```

Comment:
Also I have some small doubts that we are splitting one async IO action into two smaller ones, which can cause some extra overheads. As I understand the reason is that you want to have an existing pending checkpoint when doing error handling of `initializeCheckpointLocation`, right? Wouldn't an alternate solution to this problem be to have somewhere in the error handling logic, if we are handling an error for a checkpoint that's missing `PendingCheckpoint`, we lazily/on-demand create one? Have you considered this approach and what pros/cons would it have?

Reply:
> initializing the storage location was moved after creating pending checkpoint due to error handling reasons, right? If so, why leaving `checkpointIdCounter.getAndIncrement()` here doesn't cause the same problems?

We have handled checkpoint error, we `initialize CheckpointLocation after create PendingCheckpoint` due to we want to PendingCheckpoint as early as possible. It will show Checkpoint in CheckpointHistory Page. But we must leave  `checkpointIdCounter.getAndIncrement()` here, because checkpoint is necessary for create PendingCheckpoint.

If checkpoint throw exception before create pendingCheckpoint, it won't show checkpoint info in CheckpointHistory Page.

Reply:
> initializing the storage location was moved after creating pending checkpoint due to error handling reasons, right? If so, why leaving checkpointIdCounter.getAndIncrement() here doesn't cause the same problems?

Yes, it is correct. It will be the same problem for getAndIncrement but these changes just try to reduce the probability of this event, and also if we don't even have the checkpoint id then there is nothing to send to the statistic.

> Wouldn't an alternate solution to this problem be to have somewhere in the error handling logic, if we are handling an error for a checkpoint that's missing PendingCheckpoint, we lazily/on-demand create one?

I thought about that but I didn't find an easy way to do so. I mean right now we lose all information if we fail before the PendingCheckpoint. I also think that PendingCheckpoint is too complex an object for such fake creation. I am maybe wrong but it looks like it requires a lot of refactoring. Considering that this ticket is about logging, statistic but not about PendingCheckpoint lifecycle, I decided that we can do more deep refactoring in the other ticket if we think that it makes sense.

Reply:
@1996fanrui, maybe you have something to add to my answers?

Reply:
Hi @akalash your answer is comprehensive, I agree with do more deep refactoring in the other ticket.

Reply:
Ok, thanks for the clarifications.